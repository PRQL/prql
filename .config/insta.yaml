behavior:
  # force_update: true
review:
  # The default (true) has a small performance hit and never needed for us
  warn_undiscovered: false
  include_ignored: false
  include_hidden: false
