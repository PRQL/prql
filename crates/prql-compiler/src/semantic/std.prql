# The PRQL standard library defines the following functions and transforms.
# The definitions are whitespace insensitive, and have this form:
#
# ```
# let my_func = param1 param2 ...  -> <return_type> body_expr
# ```
#
# Where:
#   * `my_func` is the name of the function
#   * `param1` is the first parameter optionally followed by a type in "< ... >"
#   * `param2` etc. follow the same pattern as param1
#   * `<return_type>` is the type of result wrapped in "< ... >"
#   * `body_expr` defines the function body that creates the result.
#     It can be PRQL code or `internal ...` to indicate internal compiler code.

# Operators

let mul = left right -> <int || float> internal std.mul
let div_i = left right -> <int || float> internal std.div_i
let div_f = left right -> <int || float> internal std.div_f
let mod = left right -> <int || float> internal std.mod
let add = left<int || float || timestamp || date> right<int || float || timestamp || date> -> <int || float || timestamp || date> internal std.add
let sub = left<int || float || timestamp || date> right<int || float || timestamp || date> -> <int || float || timestamp || date> internal std.sub
let eq = left right -> <bool> internal std.eq
let ne = left right -> <bool> internal std.ne
let gt = left right -> <bool> internal std.gt
let lt = left right -> <bool> internal std.lt
let gte = left right -> <bool> internal std.gte
let lte = left right -> <bool> internal std.lte
let and = left<bool> right<bool> -> <bool> internal std.and
let or = left<bool> right<bool> -> <bool> internal std.or
let coalesce = left right -> internal std.coalesce
let regex_search = text pattern -> <bool> internal std.regex_search

let neg = expr<int || float> -> <int || float> internal std.neg
let not = expr<bool> -> <bool> internal std.not

# Types

## Type primitives
type int
type float
type bool
type text
type date
type time
type timestamp
type `func`

## Generic array
# TODO: an array of anything, not just nulls
type array = [null]

## Scalar
type scalar = int || float || bool || text || date || time || timestamp || null
type tuple_of_scalars = {scalar..}

## Relation (an array of tuples)
type relation = [tuple_of_scalars]

## Transform
type transform = (func relation -> relation)

# Functions

## Relational transforms
let from = func
  `default_db.source` <relation>
  -> <relation> internal from

let select = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal select

let filter = func
  condition <bool>
  tbl <relation>
  -> <relation> internal filter

let derive = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal derive

let aggregate = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal aggregate

let sort = func
  by <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal sort

let take = func
  expr <scalar>
  tbl <relation>
  -> <relation> internal take

let join = func
  `default_db.with` <relation>
  condition <bool>
  `noresolve.side`:inner
  tbl <relation>
  -> <relation> internal join

let group = func
  by<scalar || tuple_of_scalars>
  pipeline <transform>
  tbl <relation>
  -> <relation> internal group

let window = func
  rows:0..0
  range:0..0
  expanding <bool>:false
  rolling <int>:0
  pipeline <transform>
  tbl <relation>
  -> <relation> internal window

let noop = x -> x

let append = `default_db.bottom`<relation> top<relation> -> <relation> internal append
let intersect = `default_db.bottom`<relation> top<relation> -> <relation> (
  noop t = top
  join (noop b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  select t.*
)
let remove = `default_db.bottom`<relation> top<relation> -> <relation> (
  noop t = top
  join side:left (noop b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  filter (tuple_every (tuple_map _is_null b.*))
  select t.*
)
let loop = func
  pipeline <transform>
  top <relation>
  -> <relation> internal loop

## Aggregate functions
# These return either a scalar when used within `aggregate`, or a column when used anywhere else.

let min = column <array> -> <int || float || null> internal std.min

let max = column <array> -> <int || float || null> internal std.max

let sum = column <array> -> <int || float> internal std.sum

let average = column <array> -> <float || null> internal std.average

let stddev = column <array> -> <float || null> internal std.stddev

let every = column <array> -> <bool> internal std.every

let any = column <array> -> <bool> internal std.any

let concat_array = column <array> -> <text> internal std.concat_array

# Counts number of items in the column.
# Note that the count will include null values.
let count = column<array> -> <int> internal std.count

# Deprecated in favour of filterning input to the [std.count] function (not yet implemented).
@{deprecated}
let count_distinct = column <array> -> internal std.count_distinct

## Window functions
let lag =   offset <int>    column <array> -> internal std.lag
let lead =  offset <int>    column <array> -> internal std.lead
let first      = column <array> -> internal std.first
let last       = column <array> -> internal std.last
let rank       = column <array> -> internal std.rank
let rank_dense = column <array> -> internal std.rank_dense
let row_number = column <array> -> internal std.row_number

## Misc functions
let round = n_digits column -> <scalar> internal std.round
let as = `noresolve.type` column -> <scalar> internal std.as
let in = pattern value -> <bool> internal in

## Tuple functions
let tuple_every = func list -> <bool> internal tuple_every
let tuple_map = func fn <func> list -> internal tuple_map
let tuple_zip = func a b -> internal tuple_zip
let _eq = func a -> internal _eq
let _is_null = func a -> _param.a == null

## Misc
let from_text = input<text> `noresolve.format`:csv -> <relation> internal from_text

## String functions
let lower = column -> <text> internal std.lower
let upper = column -> <text> internal std.upper

## File-reading functions, primarily for DuckDB
let read_parquet = source<text> -> <relation> internal std.read_parquet
let read_csv = source<text> -> <relation> internal std.read_csv
