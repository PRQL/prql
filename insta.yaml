behavior:
  force_update: true
review:
  include_ignored: false
  include_hidden: false
  # Small performance hit and never needed for us
  warn_undiscovered: false
