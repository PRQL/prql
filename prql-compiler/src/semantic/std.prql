# Aggregate Functions
func min <scalar|column> column -> null
func max <scalar|column> column -> null
func sum <scalar|column> column -> null
func avg <scalar|column> column -> null
func stddev <scalar|column> column -> null
func average <scalar|column> column -> null
func count <scalar|column> non_null:s"*" -> null
# TODO: Possibly make this into `count distinct:true` (or like `distinct:` as an
# abbreviation of that?)
func count_distinct <scalar|column> column -> null

# Window functions
func lag<column> offset column -> null
func lead<column> offset column -> null
func first<column> offset column -> null
func last<column> offset column -> null
func rank<column> -> null
func rank_dense<column> -> null
func row_number<column> -> null

# Other functions
func round<scalar> n_digits column -> null
func as<scalar> `noresolve.type` column -> null
func in<bool> pattern value -> null

# Transform type definitions
func from<table> `default_db.source`<table> -> null
func select<table> columns<column> tbl<table> -> null
func filter<table> condition<bool> tbl<table> -> null
func derive<table> columns<column> tbl<table> -> null
func aggregate<table> a<column> tbl<table> -> null
func sort<table> by tbl<table> -> null
func take<table> expr tbl<table> -> null
func join<table> `default_db.with`<table> filter `noresolve.side`:inner tbl<table> -> null
func group<table> by pipeline tbl<table> -> null
func window<table> rows:0..0 range:0..0 expanding:false rolling:0 pipeline tbl<table> -> null

func append<table> `default_db.bottom`<table> top<table> -> null
func intersect<table> `default_db.bottom`<table> top<table> -> (
    from t = _param.top
    join b = _param.bottom (all (map _eq (zip t.* b.*)))
    select t.*
)
func remove<table> `default_db.bottom`<table> top<table> -> (
    from t = _param.top
    join side:left b = _param.bottom (all (map _eq (zip t.* b.*)))
    filter (all (map _is_null b.*))
    select t.*
)
func loop<table> pipeline top<table> -> null

# List functions
func all<bool> list<list> -> null
func map<list> fn list<list> -> null
func zip<list> a<list> b<list> -> null
func _eq<list> a<list> -> null
func _is_null a -> _param.a == null

# Misc
func from_text<table> text<string> `noresolve.format`:csv -> null
