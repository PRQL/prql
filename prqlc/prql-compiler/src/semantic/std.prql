# The PRQL standard library defines the following functions and transforms.
# The definitions are whitespace insensitive, and have this form:
#
# ```
# let my_func = param1 param2 ...  -> <return_type> body_expr
# ```
#
# Where:
#   * `my_func` is the name of the function
#   * `param1` is the first parameter optionally followed by a type in "< ... >"
#   * `param2` etc. follow the same pattern as param1
#   * `<return_type>` is the type of result wrapped in "< ... >"
#   * `body_expr` defines the function body that creates the result.
#     It can be PRQL code or `internal ...` to indicate internal compiler code.

# Operators

let mul = left right -> <int || float> internal std.mul
let div_i = left right -> <int || float> internal std.div_i
let div_f = left right -> <int || float> internal std.div_f
let mod = left right -> <int || float> internal std.mod
let add = left right -> <int || float || timestamp || date> internal std.add
let sub = left right -> <int || float || timestamp || date> internal std.sub
let eq = left right -> <bool> internal std.eq
let ne = left right -> <bool> internal std.ne
let gt = left right -> <bool> internal std.gt
let lt = left right -> <bool> internal std.lt
let gte = left right -> <bool> internal std.gte
let lte = left right -> <bool> internal std.lte
let and = left<bool> right<bool> -> <bool> internal std.and
let or = left<bool> right<bool> -> <bool> internal std.or
let coalesce = left right -> internal std.coalesce
let regex_search = text pattern -> <bool> internal std.regex_search

let neg = expr<int || float> -> <int || float> internal std.neg
let not = expr<bool> -> <bool> internal std.not

# Types

## Type primitives
type int = int
type float = float
type bool = bool
type text = text
type date = date
type time = time
type timestamp = timestamp
type `func` = func
type anytype = anytype

## Generic array
# TODO: an array of anything, not just nulls
type array = [anytype]

## Scalar
type scalar = int || float || bool || text || date || time || timestamp || null
type tuple = {anytype..}

## Range
type range = {start = scalar, end = scalar}

## Relation (an array of tuples)
type relation = [tuple]

## Transform
type transform = func relation -> relation

# Functions

## Relational transforms
let from = func
  `default_db.source` <relation>
  -> <relation> internal from

let select = func
  columns <scalar || tuple>
  tbl <relation>
  -> <relation> internal select

let filter = func
  condition <bool>
  tbl <relation>
  -> <relation> internal filter

let derive = func
  columns <scalar || tuple>
  tbl <relation>
  -> <relation> internal derive

let aggregate = func
  columns <scalar || tuple>
  tbl <relation>
  -> <relation> internal aggregate

let sort = func
  by <scalar || tuple>
  tbl <relation>
  -> <relation> internal sort

let take = func
  expr <anytype>
  tbl <relation>
  -> <relation> internal take

let join = func
  `default_db.with` <relation>
  condition <bool>
  `noresolve.side`:inner
  tbl <relation>
  -> <relation> internal join

let group = func
  by<scalar || tuple>
  pipeline <transform>
  tbl <relation>
  -> <relation> internal group

let window = func
  rows:0..0
  range:0..0
  expanding <bool>:false
  rolling <int>:0
  pipeline <transform>
  tbl <relation>
  -> <relation> internal window

let append = `default_db.bottom`<relation> top<relation> -> <relation> internal append
let intersect = `default_db.bottom`<relation> top<relation> -> <relation> (
  t = top
  join (b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  select t.*
)
let remove = `default_db.bottom`<relation> top<relation> -> <relation> (
  t = top
  join side:left (b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  filter (tuple_every (tuple_map _is_null b.*))
  select t.*
)
let loop = func
  pipeline <transform>
  top <relation>
  -> <relation> internal loop

## Aggregate functions
# These return either a scalar when used within `aggregate`, or a column when used anywhere else.

let min = column <array> -> <int || float || null> internal std.min

let max = column <array> -> <int || float || null> internal std.max

let sum = column <array> -> <int || float> internal std.sum

let average = column <array> -> <float || null> internal std.average

let stddev = column <array> -> <float || null> internal std.stddev

let all = column <array> -> <bool> internal std.all

let any = column <array> -> <bool> internal std.any

let concat_array = column <array> -> <text> internal std.concat_array

# Counts number of items in the column.
# Note that the count will include null values.
let count = column<array> -> <int> internal count

# Deprecated in favour of filterning input to the [std.count] function (not yet implemented).
@{deprecated}
let count_distinct = column <array> -> internal std.count_distinct

## Window functions
let lag =   offset <int>    column <array> -> internal std.lag
let lead =  offset <int>    column <array> -> internal std.lead
let first      = column <array> -> internal std.first
let last       = column <array> -> internal std.last
let rank       = column <array> -> internal std.rank
let rank_dense = column <array> -> internal std.rank_dense
let row_number = column <array> -> internal row_number

# Mathematical functions
module math {
  let abs = column -> <int || float> internal std.math.abs
  let floor = column -> <int> internal std.math.floor
  let ceil = column -> <int> internal std.math.ceil
  let pi = -> <float> internal std.math.pi
  let exp = column -> <int || float> internal std.math.exp
  let ln = column -> <int || float> internal std.math.ln
  let log10 = column -> <int || float> internal std.math.log10
  let log = base<int || float> column -> <int || float> internal std.math.log
  let sqrt = column -> <int || float> internal std.math.sqrt
  let degrees = column -> <int || float> internal std.math.degrees
  let radians = column -> <int || float> internal std.math.radians
  let cos = column -> <int || float> internal std.math.cos
  let acos = column -> <int || float> internal std.math.acos
  let sin = column -> <int || float> internal std.math.sin
  let asin = column -> <int || float> internal std.math.asin
  let tan = column -> <int || float> internal std.math.tan
  let atan = column -> <int || float> internal std.math.atan
  let pow = exponent<int || float> column -> <int || float> internal std.math.pow
  let round = n_digits column -> <scalar> internal std.math.round
}

## Misc functions
let as = `noresolve.type` column -> <scalar> internal std.as
let in = pattern value -> <bool> internal in

## Tuple functions
let tuple_every = func list -> <bool> internal tuple_every
let tuple_map = func fn <func> list -> internal tuple_map
let tuple_zip = func a b -> internal tuple_zip
let _eq = func a -> internal _eq
let _is_null = func a -> _param.a == null

## Misc
let from_text = input<text> `noresolve.format`:csv -> <relation> internal from_text

## Text functions
module text {
  let lower = column -> <text> internal std.text.lower
  let upper = column -> <text> internal std.text.upper
  let ltrim = column -> <text> internal std.text.ltrim
  let rtrim = column -> <text> internal std.text.rtrim
  let trim = column -> <text> internal std.text.trim
  let length = column -> <int> internal std.text.length
  let extract = offset<int> length<int> column -> <text> internal std.text.extract
  let replace = pattern<text> replacement<text> column -> <text> internal std.text.replace
  let starts_with = prefix<text> column -> <bool> internal std.text.starts_with
  let contains = substr<text> column -> <bool> internal std.text.contains
  let ends_with = suffix<text> column -> <bool> internal std.text.ends_with
}

## Date functions
module date {
  let to_text = format<text> column -> <text> internal std.date.to_text
}

## File-reading functions, primarily for DuckDB
let read_parquet = source<text> -> <relation> internal std.read_parquet
let read_csv = source<text> -> <relation> internal std.read_csv


## PRQL compiler functions

let prql_version = -> <text> internal prql_version
