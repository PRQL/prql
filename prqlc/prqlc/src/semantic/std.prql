# The PRQL standard library defines the following functions and transforms.
# The definitions are whitespace insensitive, and have this form:
#
# ```
# let my_func = param1 param2 ...  -> <return_type> body_expr
# ```
#
# Where:
#   * `my_func` is the name of the function
#   * `param1` is the first parameter optionally followed by a type in "< ... >"
#   * `param2` etc. follow the same pattern as param1
#   * `<return_type>` is the type of result wrapped in "< ... >"
#   * `body_expr` defines the function body that creates the result.
#     It can be PRQL code or `internal ...` to indicate internal compiler code.

# Operators

let mul = left right -> <int> internal std.mul
let div_i = left right -> <int> internal std.div_i
let div_f = left right -> <int> internal std.div_f
let mod = left right -> <int> internal std.mod
let add = left right -> <int> internal std.add
let sub = left right -> <int> internal std.sub
let eq = left right -> <bool> internal std.eq
let ne = left right -> <bool> internal std.ne
let gt = left right -> <bool> internal std.gt
let lt = left right -> <bool> internal std.lt
let gte = left right -> <bool> internal std.gte
let lte = left right -> <bool> internal std.lte
let and = left<bool> right<bool> -> <bool> internal std.and
let or = left<bool> right<bool> -> <bool> internal std.or
let coalesce = left right -> internal std.coalesce
let regex_search = text pattern -> <bool> internal std.regex_search

let neg = expr<int> -> <int> internal std.neg
let not = expr<bool> -> <bool> internal std.not

# Types

## Type primitives
type int = int
type float = float
type bool = bool
type text = text
type date = date
type time = time
type timestamp = timestamp
type `func` = func

## Generic array

## Range
# type range <T> = {start = T, end = T}

## Transform
# type transform <I, O> = func [I] -> [O]

# Functions

## Relational transforms
let from = func <R>
  tbl <R>
  -> <R> tbl

@(coerce_tuple 0)
@(implicit_closure 0 this:1)
let select = func <I, A>
  assigns <A>
  tbl <[I]>
  -> <[A]> internal select

@(implicit_closure 0 this:1)
let filter = func <T>
  condition <bool>
  tbl <[T]>
  -> <[T]> internal filter

@(coerce_tuple 0)
@(implicit_closure 0 this:1)
let derive = func <I, A>
  assigns <A>
  tbl <[I]>
  -> <[{I, ..A}]> internal derive

@(coerce_tuple 0)
@(implicit_closure 0 this:1)
let aggregate = func <I, A>
  assigns <A>
  tbl <[I]>
  -> <[A]> internal aggregate

@(coerce_tuple 0)
@(implicit_closure 0 this:1)
let sort = func <I>
  by
  tbl <[I]>
  -> <[I]> internal sort

let take = func <I>
  expr
  tbl <[I]>
  -> <[I]> internal take

@(implicit_closure 1 this:2 that:0)
let join = func <A, B>
  with <[B]>
  condition <bool>
  `noresolve.side`:inner
  tbl <[A]>
  -> <[{A, B}]> internal join

@(coerce_tuple 0)
@(implicit_closure 0 this:2)
let group = func <I, B, PO>
  by<B>
  pipeline <func [I - B] -> [PO]>
  tbl <[I]>
  -> <[{B, ..PO}]> internal group

let window = func
  rows:0..-1
  range:0..-1
  expanding <bool>:false
  rolling <int>:0
  pipeline
  tbl
  -> internal window

let append = bottom top -> internal append
# let intersect = bottom top -> (
#   t = top
#   join (b = bottom) (std.tuple_every (std.tuple_map std._eq (std.tuple_zip t.* b.*)))
#   select t.*
# )
# let remove = bottom top -> (
#   t = top
#   join side:left (b = bottom) (std.tuple_every (std.tuple_map std._eq (std.tuple_zip t.* b.*)))
#   filter (std.tuple_every (std.tuple_map std._is_null b.*))
#   select t.*
# )
let loop = func
  pipeline
  top
  -> internal loop

## Aggregate functions
# These return either a scalar when used within `aggregate`, or a column when used anywhere else.

let min = column -> <int> internal std.min

let max = column -> <int> internal std.max

let sum = column -> <int> internal std.sum

let average = column -> <float> internal std.average

let stddev = column -> <float> internal std.stddev

let all = column -> <bool> internal std.all

let any = column -> <bool> internal std.any

let concat_array = column -> <text> internal std.concat_array

# Counts number of items in the column.
# Note that the count will include null values.
let count = colum -> <int> internal count

# Deprecated in favour of filterning input to the [std.count] function (not yet implemented).
@{deprecated}
let count_distinct = column -> internal std.count_distinct

## Window functions
let lag =   offset <int>    column -> internal std.lag
let lead =  offset <int>    column -> internal std.lead
let first      = column -> internal std.first
let last       = column -> internal std.last
let rank       = column -> internal std.rank
let rank_dense = column -> internal std.rank_dense
let row_number = column -> internal row_number

# Mathematical functions
module math {
  let abs = column -> <int> internal std.math.abs
  let floor = column -> <int> internal std.math.floor
  let ceil = column -> <int> internal std.math.ceil
  let pi <float> = internal std.math.pi
  let exp = column -> <int> internal std.math.exp
  let ln = column -> <int> internal std.math.ln
  let log10 = column -> <int> internal std.math.log10
  let log = base<int> column -> <int> internal std.math.log
  let sqrt = column -> <int> internal std.math.sqrt
  let degrees = column -> <int> internal std.math.degrees
  let radians = column -> <int> internal std.math.radians
  let cos = column -> <int> internal std.math.cos
  let acos = column -> <int> internal std.math.acos
  let sin = column -> <int> internal std.math.sin
  let asin = column -> <int> internal std.math.asin
  let tan = column -> <int> internal std.math.tan
  let atan = column -> <int> internal std.math.atan
  let pow = exponent<int> column -> <int> internal std.math.pow
  let round = n_digits column -> internal std.math.round
}

## Misc functions
let as = `noresolve.type` column -> internal std.as
let in = pattern value -> <bool> internal in

## Tuple functions
let tuple_every = func list -> <bool> internal tuple_every
let tuple_map = func fn <func> list -> internal tuple_map
let tuple_zip = func a b -> internal tuple_zip
let _eq = func a -> internal _eq
let _is_null = func a -> a == null

## Misc
let from_text = func <R: {}> input<text> `noresolve.format`:csv -> <[R]> internal from_text

## Text functions
module text {
  let lower = column -> <text> internal std.text.lower
  let upper = column -> <text> internal std.text.upper
  let ltrim = column -> <text> internal std.text.ltrim
  let rtrim = column -> <text> internal std.text.rtrim
  let trim = column -> <text> internal std.text.trim
  let length = column -> <int> internal std.text.length
  let extract = offset<int> length<int> column -> <text> internal std.text.extract
  let replace = pattern<text> replacement<text> column -> <text> internal std.text.replace
  let starts_with = prefix<text> column -> <bool> internal std.text.starts_with
  let contains = substr<text> column -> <bool> internal std.text.contains
  let ends_with = suffix<text> column -> <bool> internal std.text.ends_with
}

## Date functions
module date {
  let to_text = format<text> column -> <text> internal std.date.to_text
}

## File-reading functions, primarily for DuckDB
let read_parquet = func <R: {}> source<text> -> <[R]> internal std.read_parquet
# let read_csv = func <R: {}> source<text> -> <[R]> internal std.read_csv

## PRQL compiler functions
module `prql` {
  let version <text> = internal prql_version
}

# Deprecated, will be removed in 0.12.0
let prql_version <text> = internal prql_version
