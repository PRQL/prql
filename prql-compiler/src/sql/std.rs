use std::collections::HashMap;
use std::iter::zip;

use anyhow::Result;
use once_cell::sync::Lazy;
use sqlparser::ast::{self as sql_ast};

use super::gen_expr::translate_sstring;
use super::Context;
use crate::ast::{pl, rq};
use crate::semantic;

static STD: Lazy<semantic::Module> = Lazy::new(load_std_impl);

fn load_std_impl() -> semantic::Module {
    use crate::parser::parse;
    let std_lib = include_str!("./std_impl.prql");
    let statements = parse(std_lib).unwrap();

    let context = semantic::Context {
        root_mod: semantic::Module::new(),
        ..semantic::Context::default()
    };

    let (_, context) = semantic::resolve_only(statements, Some(context)).unwrap();
    let std = context.root_mod.get(&pl::Ident::from_name("std")).unwrap();

    std.kind.clone().into_module().unwrap()
}

pub(super) fn translate_built_in(
    name: String,
    args: Vec<rq::Expr>,
    ctx: &mut Context,
) -> Result<sql_ast::Expr> {
    let name = name.strip_prefix("std.").unwrap();

    let entry = STD.get(&pl::Ident::from_name(name)).unwrap();
    let func_def = entry.kind.as_func_def().unwrap();

    let params = func_def
        .named_params
        .iter()
        .chain(func_def.positional_params.iter())
        .map(|x| x.name.split('.').last().unwrap_or(x.name.as_str()));

    let mut args: HashMap<&str, _> = zip(params, args.into_iter()).collect();

    // body can only be an s-string
    let body = &func_def.body.kind.as_s_string().unwrap();
    let body = body
        .iter()
        .map(|item| {
            match item {
                pl::InterpolateItem::Expr(expr) => {
                    // s-string exprs can only contain idents
                    let ident = expr.kind.as_ident();
                    let ident = ident.as_ref().unwrap();

                    // lookup args
                    let arg = args.remove(ident.name.as_str());
                    pl::InterpolateItem::<rq::Expr>::Expr(Box::new(arg.unwrap()))
                }
                pl::InterpolateItem::String(s) => pl::InterpolateItem::String(s.clone()),
            }
        })
        .collect::<Vec<_>>();

    let s_string = translate_sstring(body, ctx)?;

    Ok(sql_ast::Expr::Identifier(sql_ast::Ident::new(s_string)))
}
