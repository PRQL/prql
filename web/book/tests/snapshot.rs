#![cfg(not(target_family = "wasm"))]
use anyhow::{bail, Result};
use globset::Glob;
use insta::assert_snapshot;
use mdbook_prql::code_block_lang_tags;
use prql_compiler::*;
use std::path::{Path, PathBuf};
use std::{collections::HashMap, fs};
use walkdir::WalkDir;

#[test]
/// This test:
/// - Extracts PRQL code blocks from the book
/// - Compiles them to SQL, comparing to a snapshot. Insta raises an error if
///   there's a diff.
///
/// This mirrors the process in [replace_examples], which inserts a
/// comparison table of SQL into the book, and so serves as a snapshot test of
/// those examples.
/// Snapshot the SQL output of each example.
fn test_prql_examples() {
    let opts = Options::default().no_signature();
    let examples = collect_book_examples().unwrap();

    for (path, prql) in examples {
        // Whether it's a success or a failure, get the string.
        let sql = compile(&prql, &opts).unwrap_or_else(|e| e.to_string());
        assert_snapshot!(path.to_str().unwrap(), &sql, &prql);
    }
}

const ROOT_EXAMPLES_PATH: &str = "tests/prql";

/// Collect all the PRQL examples in the book, as a map of <Path, PRQL>.
fn collect_book_examples() -> Result<HashMap<PathBuf, String>> {
    // TODO: instead of returning Strings with embedded tags (e.g. `# error`),
    // we could instead return a struct with a `prql` field and a struct of its
    // metadata. That would make `test_display` work by matching on the metadata
    // rather than re-parsing the string.
    use pulldown_cmark::{Event, Parser};
    let glob = Glob::new("**/*.md")?.compile_matcher();
    let examples_in_book: HashMap<PathBuf, String> = WalkDir::new(Path::new("./src/"))
        .into_iter()
        .flatten()
        .filter(|x| glob.is_match(x.path()))
        .flat_map(|dir_entry| {
            let text = fs::read_to_string(dir_entry.path())?;
            // TODO: Still slightly duplicative logic here and in
            // [lib.rs/replace_examples], but not sure how to avoid it.
            //
            let mut parser = Parser::new(&text);
            let mut prql_blocks = vec![];
            while let Some(event) = parser.next() {
                let Some(lang_tags) = code_block_lang_tags(&event) else {
                    continue
                };

                if lang_tags.contains(&"prql".to_string()) {
                    let mut prql_text = String::new();
                    while let Some(Event::Text(line)) = parser.next() {
                        prql_text.push_str(line.to_string().as_str());
                    }
                    if prql_text.is_empty() {
                        bail!("Expected text after PRQL code block");
                    }
                    if lang_tags.contains(&"error".to_string()) {
                        prql_blocks.push(format!("# error\n\n{prql_text}"));
                    } else if lang_tags.contains(&"no-fmt".to_string()) {
                        prql_blocks.push(format!("# no-fmt\n\n{prql_text}"));
                    } else {
                        prql_blocks.push(prql_text.to_string());
                    }
                }
            }
            let snapshot_prefix = &dir_entry
                .path()
                .strip_prefix("./src/")?
                .to_str()
                .unwrap()
                .trim_end_matches(".md");
            Ok(prql_blocks
                .iter()
                .enumerate()
                .map(|(i, example)| {
                    (
                        Path::new(&format!("{ROOT_EXAMPLES_PATH}/{snapshot_prefix}-{i}.prql"))
                            .to_path_buf(),
                        example.to_string(),
                    )
                })
                .collect::<HashMap<_, _>>())
        })
        .flatten()
        .collect();

    Ok(examples_in_book)
}

/// Test that the formatted result (the `Display` result) of each example can be
/// compiled.
//
// We previously snapshot all the queries. But that was a lot of output, for
// something we weren't yet looking at.
//
// The ideal would be to auto-format the examples themselves, likely during the
// compilation. For that to provide a good output, we need to implement a proper
// autoformatter.
#[test]
fn test_display() -> Result<(), ErrorMessages> {
    collect_book_examples()?
        .iter()
        .try_for_each(|(path, prql)| {
            if prql.contains("# error") || prql.contains("# no-fmt") {
                return Ok(());
            }
            prql_to_pl(prql)
                .and_then(pl_to_prql)
                .and_then(|formatted| compile(&formatted, &Options::default()))
                .unwrap_or_else(|_| {
                    panic!(
                        "
Failed compiling the formatted result of {path:?}
To skip this test for an example, use `prql,no-fmt` as the language label.

The original PRQL was:

{prql}

",
                        path = path.canonicalize().unwrap(),
                        prql = prql
                    )
                });

            Ok::<(), ErrorMessages>(())
        })?;

    Ok(())
}

#[test]
fn test_rq_serialize() -> Result<(), ErrorMessages> {
    for (_, prql) in collect_book_examples()? {
        if prql.contains("# error") {
            continue;
        }
        let rq = prql_to_pl(&prql).map(pl_to_rq)?;
        // Serialize to YAML
        assert!(serde_yaml::to_string(&rq).is_ok());
    }

    Ok(())
}
