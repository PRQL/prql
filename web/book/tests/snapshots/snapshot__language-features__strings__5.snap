---
source: web/book/tests/snapshot.rs
expression: "from my_table\nselect x = \"\\t\\tline ends here\\n \\\\ \"\n"
---
SELECT
  '		line ends here
 \ ' AS x
FROM
  my_table

